### PR TITLE
Update tfe architecture

### PR DESCRIPTION
**Pull request by ml4**
- [ ] inaccuracy
- [x] clarification
- [x] new docs
- [ ] cosmetic bug - fixing broken text or markup
- [ ] enhancement - changing the website's behavior/layout

Merging the 2019 TFE internal architecture gdoc into public docs to help customers with increasingly deep and probing questions as adoption rises.

### CODE REVIEW

**Comment by sudomateo** [comment]
Going to run this locally and read through it.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (amy-hashi) [inline content/source/docs/enterprise/system-overview/architecture.html.md:186]: support-bundle runs in a/a when called, same as standalone.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (amy-hashi) [inline content/source/docs/enterprise/system-overview/architecture.html.md:157]: so, this is all the long-running service containers, but not every container. there are a variety of ephemeral containers that run on startup and for batch jobs that will show up with random names.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (amy-hashi) [inline content/source/docs/enterprise/system-overview/architecture.html.md:109]: PTFE - the P is silent..

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (amy-hashi) [inline content/source/docs/enterprise/system-overview/architecture.html.md:8]: We generally avoid abbreviating the product name anywhere on our website where we can avoid it. TFE should be spelled out as Terraform Enterprise throughout out website documentation.

**Review by amy-hashi — COMMENTED** [review]
I've got some specific comments, but also, I think we'd like to hold off on publishing this as there are some re-names in flight as well as some architecture changes coming in the semi-near future.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (aaron-lane) [inline content/source/docs/enterprise/system-overview/architecture.html.md:8]: Capitalization fix:
```suggestion
Terraform Enterprise (TFE) is powered by the same set of services behind HashiCorp’s public SaaS application Terraform Cloud. The sections below present the functional layers of the architecture, how they interact and details pertaining to the corresponding Docker containers involved. References to Docker containers in this document are those you will see when running a `docker ps` on a TFE application server while the application is up.
```

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (aaron-lane) [inline content/source/docs/enterprise/system-overview/architecture.html.md:12]: Wording suggestion:
```suggestion
~> **Note:** Several services described below include the legacy string `ptfe` as part of their names. See [here](https://www.terraform.io/docs/enterprise/index.html#note-about-product-names) for more details.
```

**Review by aaron-lane — CHANGES_REQUESTED** [review]
A couple of small suggestions follow. I don't want to overdo the review since a couple other team members are already on it.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:49]: Content updated

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:45]: Content updated

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:131]: I got this content from the original gdoc and am effectively just reformatting it for publication. I redrew the diagrams for the same reason but they come exactly from that doc - step 8 cited corresponds to a point in the image above it so if amendments are required to the other diagrams correspondingly, could you tell me exactly where to make the changes please?

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:157]: Is it the view of engineering that these non-randomly-named but ephemeral containers should be included in this internal doc?

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (aaron-lane) [inline content/source/docs/enterprise/system-overview/architecture.html.md:49]: @amy-hashi What do you think about this suggestion to help clarify this section? I removed the mention of Redis because it is already discussed in the `Coordination Layer` section and the sentence seemed to imply that Redis was also part of the storage layer.

```suggestion
All application data is stored in a PostgreSQL database and an S3-compatible object storage service. If you have selected Mounted Disk mode, local services will be started for you; otherwise, the application must be configured to access an external database and object storage service.
```

**Review by aaron-lane — APPROVED** [review]
Thanks for incorporating my suggestions, @ml4.

LGTM 📜

**Comment by ml4** [comment]
Hello all, we're getting a little compression on internals from banks - would it be OK to merge it as it and update with ephemeral container list and update container names later please?

**Comment by laurapacilio** [comment]
Hi all - wanted to check in on this one, too! Is this something we'd want to merge in still/would we still want to make a few changes?

**Comment by aaron-lane** [comment]
@amy-hashi @sudomateo ☝🏼

**Comment by ml4** [comment]
There is still an increased level of interest primarily from global banks in knowing the internals as per the PR so that they can maintain decent comms with their security and compliance teams.

**Comment by laurapacilio** [comment]
Nice okay - I'm going to look through then to just do a quick review for flow, grammar, etc. @amy-hashi and @sudomateo would you mind also giving it a final thumbs up as well? Then we can get it merged in!

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:25]:
```suggestion
A web service receives a request and enqueues a background job to perform the task via the Coordination Layer (see below). The web service then responds immediately to the original request instead of waiting for the job to be finished.
```

Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:25]: I might make these into numbered steps, since they're a progression and move them above the image.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:55]: Are these the Docker containers? If so maybe we mention that?

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:57]: In all of these sections, I'd move the steps to be above the diagram. I think that will be more cohesive for users (to set them up for what they're observing in the diagram, and then we also won't split sentences that say "the following steps" from the following steps.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:5]: Could you add a metadata description to the top of this page? 

description: |- 
  Learn about the functional layers of the architecture, how they interact, and details about the corresponding Docker containers.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:78]:
```suggestion
1. Terraform launches the remaining services that do not rely on PostgreSQL in parallel. These include two short-lived containers, `ptfe_migrations` and `ptfe_registry_migrations`, that exist only to apply any necessary updates to the database schema.<br />
```

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:87]: Question (newbie here!) this "the private module registry," indicating that there's a universal private module registry. Is that the case, or is this "your private module registry."

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:97]:
```suggestion
At this point, each tagged version of your repository has been persisted in the application. The process to make the modules available in your private module registry begins:
```

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (sudomateo) [inline content/source/docs/enterprise/system-overview/architecture.html.md:12]: This is not technically a "legacy" string if there hasn't been an deprecation notice about replacing it. As much as I too want it gone.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (sudomateo) [inline content/source/docs/enterprise/system-overview/architecture.html.md:18]:
```suggestion
- Web Services - Answer requests forwarded by Nginx and enqueue jobs for later
```

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:109]: Just changing to active voice because it's easier to read - but please reject if this changes the meaning too much!

```suggestion
Communications between Terraform Enterprise and a connected VCS server happen via HTTP webhooks. Terraform downloads the latest state of the connected Git repository using the process below, and then the state machine applies the followup Terraform commands.
```

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (sudomateo) [reply content/source/docs/enterprise/system-overview/architecture.html.md:49]: +1 to this suggestion.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (sudomateo) [inline content/source/docs/enterprise/system-overview/architecture.html.md:51]: Not sure what "outside" really means here.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (aaron-lane) [reply content/source/docs/enterprise/system-overview/architecture.html.md:49]: @ml4 What do you think?

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (sudomateo) [inline content/source/docs/enterprise/system-overview/architecture.html.md:83]: I believe we should remove this section entirely as this is subject to change and actually changing as we speak.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:157]:
```suggestion
The following table lists every container Terraform Enterprise runs via the Replicated Native Scheduler, following the layers outlined above. It does not include ephemeral containers, only those from long-running processes.
```

**Review by sudomateo — CHANGES_REQUESTED** [review]
I am not in favor of posting the container names. Instead, I would personally use the service names and how they interact with each other. For example, instead of saying that `ptfe_atlas` requests a request from `ptfe_nginx`, I would say the Atlas service receives a request from Nginx.

The container names may change over time, but the service names won't. Plus, in mounted disk mode many containers have `_disk` appended to the container name, making this document partially correct.

I also am not in favor of adding the long running container list to this page. While I think we should have more transparency around the containers that execute as part of Terraform Enterprise, I do not believe this page is the correct location for that content for reasons I mentioned above and because there are some ephemeral containers that run that do not have a name and are not listed here.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:166]:
```suggestion
|  | `ptfe_build_manager` | Terraform Build Manager | Go service that coordinates Terraform runs, plans, and applies. |
```

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:167]: I think Terraform Registry is a proper noun, but this isn't. Please correct if if I'm wrong!

```suggestion
|  | `ptfe_registry_worker` | Private Module Registry Worker | Go service that parses Git repositories for private registry. |
```

**Review by laurapacilio — CHANGES_REQUESTED** [review]

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:14]: I'll pull `legacy`

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:53]: I was separating Replicated from TFE here but I think your call is perspicacious, so I'll pull the comment.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:85]: What is it changing to and when will the release happen? I'll comment `Wait for Dependencies`.

File: content/source/docs/enterprise/system-overview/architecture.html.md
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:89]: PMR is organization-centric.

**Comment by ml4** [comment]
> I am not in favor of posting the container names. Instead, I would personally use the service names and how they interact with each other. For example, instead of saying that `ptfe_atlas` requests a request from `ptfe_nginx`, I would say the Atlas service receives a request from Nginx.
> 
> The container names may change over time, but the service names won't. Plus, in mounted disk mode many containers have `_disk` appended to the container name, making this document partially correct.
> 
> I also am not in favor of adding the long running container list to this page. While I think we should have more transparency around the containers that execute as part of Terraform Enterprise, I do not believe this page is the correct location for that content for reasons I mentioned above and because there are some ephemeral containers that run that do not have a name and are not listed here.
> 
> TL;DR Since this is an architecture overview page, I believe the content should stick with high-level service names instead of container names.

We are getting *explicit* requests from the field for a full list of container names as customers see them in a `docker ps` output and corresponding functions.  I have had another one from another global bank in the last day, and I'm sensing more than a degree of dissatisfaction about it, as technical teams feel they are somewhat less able to explain specific functionality to their compliance/security/audit teams about how the platform works due to the perceived paucity of documentation of this level.  

If we are to level up to our customer's expectations we need to absolutely list interactions as `ptfe_atlas` requests a request from `ptfe_nginx` _instead_ of Atlas service receives a request from Nginx, augmenting the documentation as necessary to ensure customers know that `ptfe_nginx` is part of, or the whole of, the Nginx service as necessary.  We then also need to list all of the containers that form part of the platform, and update them accordingly during releases in the same way we do other areas of the documentation.

I originally began this work to address customer concerns and used @glasner 's original document as a starting point as it had been heavily commented by Engineering in the regard so presumed it had pretty much already been agreed on.  With the deepest possible respect, I would propose we address the content so that it is not deemed partially correct but instead wholly correct, migrating areas of the content to other pages as needed and make it live as soon as possible.

**Comment by ihuntenator** [comment]
As a customer I really prefer this page here: https://github.com/hashicorp/terraform-website/blob/updateTFEarchitecture/content/source/docs/enterprise/system-overview/architecture.html.md
compared to what is currently here: https://www.terraform.io/docs/enterprise/system-overview/architecture.html

**Pull request closed without merging**
